src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/AggregationManager.cs

```diff
@@ -0,0 +1,314 @@
+            return this;
+        }
+
+        public AggregationManager Include(Predicate<Instrument> instrumentFilter)
+        {
+            lock (this)
+            {
+                _instrumentConfigFuncs.Add(instrumentFilter);
+            }
+            return this;
+        }
+
+        public AggregationManager SetCollectionPeriod(TimeSpan collectionPeriod)
+        {
+            lock (this)
+            {
+                _collectionPeriod = collectionPeriod;
+            }
+            return this;
+        }
+
+        public void Start()
+        {
+            // if already started or already stopped we can't be started again
+            if (_collectThread != null || _cts.IsCancellationRequested)
+            {
+                // this is purely for defensive programming and it should be impossible
+                // for a developer to observe this exception unless this library has a
+                // bug in it. Because of that we didn't bother to localize the error message.
+                throw new InvalidOperationException("Start can only be called once");
+            }
+
+            if (_collectionPeriod.TotalSeconds < MinCollectionTimeSecs)
+            {
+                // this is purely for defensive programming and it should be impossible
+                // for a developer to observe this exception unless this library has a
+                // bug in it. Because of that we didn't bother to localize the error message.
+                throw new InvalidOperationException($"CollectionPeriod must be >= {MinCollectionTimeSecs} sec");
+            }
+
+            // This explicitly uses a Thread and not a Task so that metrics still work
+            // even when an app is experiencing thread-pool starvation. Although we
+            // can't make in-proc metrics robust to everything, this is a common enough
+            // problem in .NET apps that it feels worthwhile to take the precaution.
+            _collectThread = new Thread(() => CollectWorker(_cts.Token));
+            _collectThread.Start();
+
+            _listener.Start();
+            _initialInstrumentEnumerationComplete();
+        }
+
+        private void CollectWorker(CancellationToken cancelToken)
+        {
+            try
+            {
+                double collectionIntervalSecs = -1;
+                lock (this)
+                {
+                    collectionIntervalSecs = _collectionPeriod.TotalSeconds;
+                }
+                if (collectionIntervalSecs < MinCollectionTimeSecs)
+                {
+                    // this is purely for defensive programming and it should be impossible
+                    // for a developer to observe this exception unless this library has a
+                    // bug in it. Because of that we didn't bother to localize the error message.
+                    throw new InvalidOperationException($"_collectionPeriod must be >= {MinCollectionTimeSecs} sec");
```

Comment:
Same comment.  If the only way to hit this is because we have a bug, it should be an assert.